src/coreclr/vm/ceemain.cpp

```diff
@@ -1702,6 +1702,130 @@ BOOL STDMETHODCALLTYPE EEDllMain( // TRUE on success, FALSE on error.
 
 #endif // !defined(CORECLR_EMBEDDED)
 
+static void RuntimeThreadShutdown(void* thread)
+{
+    Thread* pThread = (Thread*)thread;
+    _ASSERTE(pThread == GetThreadNULLOk());
+
+    if (pThread)
+    {
+#ifdef FEATURE_COMINTEROP
+        // reset the CoInitialize state
+        // so we don't call CoUninitialize during thread detach
+        pThread->ResetCoInitialized();
+#endif // FEATURE_COMINTEROP
+        // For case where thread calls ExitThread directly, we need to reset the
+        // frame pointer. Otherwise stackwalk would AV. We need to do it in cooperative mode.
+        // We need to set m_GCOnTransitionsOK so this thread won't trigger GC when toggle GC mode
+        if (pThread->m_pFrame != FRAME_TOP)
+        {
+#ifdef _DEBUG
+            pThread->m_GCOnTransitionsOK = FALSE;
+#endif
+            GCX_COOP_NO_DTOR();
+            pThread->m_pFrame = FRAME_TOP;
+            GCX_COOP_NO_DTOR_END();
+        }
+
+        pThread->DetachThread(TRUE);
+    }
+    else
+    {
+        // Since we don't actually cleanup the TLS data along this path, verify that it is already cleaned up
+        AssertThreadStaticDataFreed();
+    }
+
+    ThreadDetaching();
+}
+
+#ifdef TARGET_WINDOWS
+
+// Index for the fiber local storage of the attached thread pointer
+static uint32_t g_flsIndex = FLS_OUT_OF_INDEXES;
+
+// This is called when each *fiber* is destroyed. When the home fiber of a thread is destroyed,
+// it means that the thread itself is destroyed.
+// Since we receive that notification outside of the Loader Lock, it allows us to safely acquire
+// the ThreadStore lock in the RuntimeThreadShutdown.
+static void __stdcall FiberDetachCallback(void* lpFlsData)
+{
+    ASSERT(g_flsIndex != FLS_OUT_OF_INDEXES);
+    ASSERT(lpFlsData == FlsGetValue(g_flsIndex));
+
+    if (lpFlsData != NULL)
+    {
+        // The current fiber is the home fiber of a thread, so the thread is shutting down
+        RuntimeThreadShutdown(lpFlsData);
+    }
+}
+
+bool InitFlsSlot()
+{
+    // We use fiber detach callbacks to run our thread shutdown code because the fiber detach
+    // callback is made without the OS loader lock
+    g_flsIndex = FlsAlloc(FiberDetachCallback);
+    if (g_flsIndex == FLS_OUT_OF_INDEXES)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+// Register the thread with OS to be notified when thread is about to be destroyed
+// It fails fast if a different thread was already registered with the current fiber.
+// Parameters:
+//  thread        - thread to attach
+static void OsAttachThread(void* thread)
+{
+    void* threadFromCurrentFiber = FlsGetValue(g_flsIndex);
+
+    if (threadFromCurrentFiber != NULL)
+    {
+        _ASSERTE(!"Multiple threads encountered from a single fiber");
+        COMPlusThrowWin32();
```

Comment:
This should be fail fast or `ASSERTE_ALL_BUILDS`. I do think we want to run any exception handling here.

Reply:
`ASSERTE_ALL_BUILDS` would be the right choice if it did what the name implies, but is it just a regular assert?

Reply:
Actually it looks like `ASSERTE_ALL_BUILDS` _is_ a failfast in CoreClr. It is in the NativeAOT where it is just a regular debug-only assert.

Reply:
I saw the pattern commonly used in threeads.cpp. So I figured it might be the convention for this kind of code.

https://github.com/dotnet/runtime/blob/ca5cf417e37028812aa40111951b31f8808b3396/src/coreclr/vm/threads.cpp#L1051-L1055

https://github.com/dotnet/runtime/blob/ca5cf417e37028812aa40111951b31f8808b3396/src/coreclr/vm/threads.cpp#L1138-L1139

```
. . . bunch of other cases . . .
```

I do agree that ASSERTE_ALL_BUILDS intuitively seems more appropriate.

Reply:
It is a convention to throw an exception when there is a (transient) error during runtime initialization.

This case is different: It means that something is really wrong, e.g. the process uses fibers that .NET runtime does not support.

Reply:
Right. In this case these are situations that we cannot recover from, so should be a failfast or an equivalent.